File: clients/dart/fylo.dart
// Fylo client — drives the `fylo` binary's persistent NDJSON loop.
//
// Dart SDK only (dart:io, dart:convert, dart:async). Requires the `fylo` binary
// on PATH (brew/scoop) or an explicit path. One long-lived subprocess keeps the
// engine warm across calls.
//
//   final db = await Fylo.open('/path/to/db');
//   await db.createCollection('users');
//   final id = await db.putData('users', {'name': 'Ada', 'role': 'admin'});
//   final doc = await db.getLatest('users', id);
//   final admins = await db.findDocs('users', {r'$ops': [{'role': {r'$eq': 'admin'}}]});
//   await db.close();
//
// Each operation method builds the request and resolves with the op's `result`
// (throwing on failure). Method names mirror the machine-protocol op names in
// Dart's lowerCamelCase. `request(op)` remains a raw escape hatch resolving the
// full response — use it for ops without a dedicated method (branching, schema…).
// Requests are queued: each resolves with its own response line, in order.

import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

class FyloException implements Exception {
  final String message;
  FyloException(this.message);
  @override
  String toString() => 'FyloException: $message';
}

/// Metadata used with [Fylo.queueConsumer]. Dart does not provide runtime
/// decorator execution, so annotated handlers are explicitly bound to a Fylo.
class FyloQueueConsumer {
  final String topic;
  final String group;
  final int maxMessages;
  final int visibilityTimeoutMs;
  final int maxAttempts;
  final int retryDelayMs;

  const FyloQueueConsumer(this.topic, this.group,
      {this.maxMessages = 1,
      this.visibilityTimeoutMs = 30000,
      this.maxAttempts = 3,
      this.retryDelayMs = 0});
}

class QueueProcessResult {
  final int claimed;
  int acknowledged = 0;
  int retried = 0;
  int deadLettered = 0;

  QueueProcessResult(this.claimed);
}

typedef QueueHandler = FutureOr<void> Function(Map<String, dynamic> delivery);
typedef QueueRunner = Future<QueueProcessResult> Function();

class Fylo {
  static const int maxRequestBytes = 1024 * 1024;
  static const int maxResponseBytes = 8 * 1024 * 1024;
  final Process _proc;
  final IOSink _stdin;
  final List<Completer<Map<String, dynamic>>> _queue = [];
  final Uint8List _responseBuffer = Uint8List(maxResponseBytes);
  int _responseLength = 0;
  bool _responseOversized = false;

  Fylo._(this._proc, this._stdin, Stream<List<int>> bytes) {
    bytes.listen((chunk) {
      for (final byte in chunk) {
        if (byte != 0x0a) {
          if (!_responseOversized) {
            if (_responseLength >= maxResponseBytes) {
              _responseLength = 0;
              _responseOversized = true;
            } else {
              _responseBuffer[_responseLength++] = byte;
            }
          }
          continue;
        }
        if (_responseOversized) {
          _failAll(
              FyloException('FYLO response exceeds $maxResponseBytes bytes'));
          _proc.kill();
          return;
        }
        if (_queue.isNotEmpty) {
          final completer = _queue.removeAt(0);
          try {
            final line = utf8.decode(
                Uint8List.sublistView(_responseBuffer, 0, _responseLength),
                allowMalformed: false);
            completer.complete(jsonDecode(line) as Map<String, dynamic>);
          } catch (error) {
            completer.completeError(
                FyloException('fylo returned malformed UTF-8 or JSON'));
            _failAll(FyloException('fylo returned malformed UTF-8 or JSON'));
            _proc.kill();
            return;
          }
        }
        _responseLength = 0;
        _responseOversized = false;
      }
    }, onDone: () {
      _failAll(FyloException('fylo process exited'));
    });
  }

  void _failAll(FyloException error) {
    for (final completer in _queue) {
      if (!completer.isCompleted) completer.completeError(error);
    }
    _queue.clear();
  }

  /// Start a warm fylo process rooted at [root]. [binary] defaults to `fylo`.
  static Future<Fylo> open(String root, {String binary = 'fylo'}) async {
    final args = [
      'exec',
      '--loop',
      '--root',
      root,
      '--max-request-bytes',
      '$maxRequestBytes',
      '--max-response-bytes',
      '$maxResponseBytes'
    ];
    final proc = await Process.start(binary, args);
    return Fylo._(proc, proc.stdin, proc.stdout);
  }

  /// Send one raw machine-protocol op; resolves with the full response object.
  Future<Map<String, dynamic>> request(Map<String, dynamic> op) {
    final payload = utf8.encode(jsonEncode(op));
    if (payload.length > maxRequestBytes) {
      return Future.error(
          FyloException('FYLO request exceeds $maxRequestBytes bytes'));
    }
    final completer = Completer<Map<String, dynamic>>();
    _queue.add(completer);
    _stdin.add([...payload, 0x0a]);
    return completer.future;
  }

  Future<dynamic> _op(String name, Map<String, dynamic> fields) async {
    final payload = <String, dynamic>{'op': name};
    fields.forEach((key, value) {
      if (value != null) payload[key] = value;
    });
    final response = await request(payload);
    if (response['ok'] != true) {
      final error = response['error'];
      throw FyloException(
          (error is Map ? error['message'] : null) ?? 'fylo error');
    }
    return response['result'];
  }

  // --- Collections ---
  Future<dynamic> createCollection(String collection,
          [String kind = 'document']) =>
      _op('createCollection', {'collection': collection, 'kind': kind});
  Future<dynamic> dropCollection(String collection) =>
      _op('dropCollection', {'collection': collection});
  Future<dynamic> inspectCollection(String collection) =>
      _op('inspectCollection', {'collection': collection});
  Future<dynamic> rebuildCollection(String collection) =>
      _op('rebuildCollection', {'collection': collection});

  // --- Durable serverless queue ---
  Future<dynamic> queuePublish(String topic, dynamic payload,
          [Map<String, dynamic> options = const {}]) =>
      _op('queuePublish', {
        'topic': topic,
        'payload': payload,
        ..._queueOptions(options, const {'delayMs', 'idempotencyKey'})
      });
  Future<dynamic> queueClaim(String topic, String group,
          [Map<String, dynamic> options = const {}]) =>
      _op('queueClaim', {
        'topic': topic,
        'group': group,
        ..._queueOptions(options,
            const {'maxMessages', 'visibilityTimeoutMs', 'maxAttempts'})
      });
  Future<dynamic> queueAck(
          String topic, String group, String id, String receipt) =>
      _op('queueAck',
          {'topic': topic, 'group': group, 'id': id, 'receipt': receipt});
  Future<dynamic> queueNack(
          String topic, String group, String id, String receipt,
          [Map<String, dynamic> options = const {}]) =>
      _op('queueNack', {
        'topic': topic,
        'group': group,
        'id': id,
        'receipt': receipt,
        ..._queueOptions(options, const {'delayMs', 'reason'})
      });
  Future<dynamic> queueExtend(String topic, String group, String id,
          String receipt, int visibilityTimeoutMs) =>
      _op('queueExtend', {
        'topic': topic,
        'group': group,
        'id': id,
        'receipt': receipt,
        'visibilityTimeoutMs': visibilityTimeoutMs
      });
  Future<dynamic> queueStats(String topic, String group) =>
      _op('queueStats', {'topic': topic, 'group': group});
  Future<dynamic> queueDeadLetters(String topic, String group,
          [int limit = 100]) =>
      _op('queueDeadLetters', {'topic': topic, 'group': group, 'limit': limit});

  /// Copy documented queue options only; unknown and protected fields are ignored.
  static Map<String, dynamic> _queueOptions(
          Map<String, dynamic> options, Set<String> allowed) =>
      {
        for (final key in allowed)
          if (options.containsKey(key)) key: options[key]
      };

  /// Process and settle one bounded queue batch.
  Future<QueueProcessResult> queueProcess(
      String topic, String group, QueueHandler handler,
      [FyloQueueConsumer? options]) async {
    final config = options ?? FyloQueueConsumer(topic, group);
    final claimed = await queueClaim(topic, group, {
      'maxMessages': config.maxMessages,
      'visibilityTimeoutMs': config.visibilityTimeoutMs,
      'maxAttempts': config.maxAttempts
    });
    if (claimed is! List) {
      throw FyloException('fylo queue claim returned an invalid delivery list');
    }
    final result = QueueProcessResult(claimed.length);
    for (final item in claimed) {
      if (item is! Map) {
        throw FyloException('fylo queue claim returned an invalid delivery');
      }
      final delivery = Map<String, dynamic>.from(item);
      final id = delivery['id'];
      final receipt = delivery['receipt'];
      if (id is! String || receipt is! String) {
        throw FyloException('queue delivery lacks an id or receipt');
      }
      var failed = false;
      try {
        await handler(delivery);
      } catch (_) {
        failed = true;
      }
      if (!failed) {
        await queueAck(topic, group, id, receipt);
        result.acknowledged++;
      } else {
        final settled = await queueNack(topic, group, id, receipt,
            {'delayMs': config.retryDelayMs, 'reason': 'queue handler failed'});
        if (settled is Map && settled['deadLettered'] == true) {
          result.deadLettered++;
        } else {
          result.retried++;
        }
      }
    }
    return result;
  }

  /// Bind annotated metadata and a handler into a one-batch invocation.
  QueueRunner queueConsumer(FyloQueueConsumer consumer, QueueHandler handler) =>
      () => queueProcess(consumer.topic, consumer.group, handler, consumer);

  // --- Documents ---
  Future<dynamic> putData(String collection, Map<String, dynamic> data) =>
      _op('putData', {'collection': collection, 'data': data});
  Future<dynamic> batchPutData(String collection, List<dynamic> batch) =>
      _op('batchPutData', {'collection': collection, 'batch': batch});
  Future<dynamic> getDoc(String collection, String id) =>
      _op('getDoc', {'collection': collection, 'id': id});
  Future<dynamic> getMeta(String collection, String id) =>
      _op('getMeta', {'collection': collection, 'id': id});
  Future<dynamic> setMeta(
          String collection, String id, Map<String, dynamic> meta) =>
      _op('setMeta', {'collection': collection, 'id': id, 'meta': meta});
  Future<dynamic> getLatest(String collection, String id) =>
      _op('getLatest', {'collection': collection, 'id': id});
  Future<dynamic> patchDoc(
          String collection, String id, Map<String, dynamic> newDoc) =>
      _op('patchDoc', {'collection': collection, 'id': id, 'newDoc': newDoc});
  Future<dynamic> patchDocs(String collection, Map<String, dynamic> update) =>
      _op('patchDocs', {'collection': collection, 'update': update});
  Future<dynamic> delDoc(String collection, String id) =>
      _op('delDoc', {'collection': collection, 'id': id});
  Future<dynamic> delDocs(String collection, Map<String, dynamic> criteria) =>
      _op('delDocs', {'collection': collection, 'delete': criteria});
  Future<dynamic> restoreDoc(String collection, String id) =>
      _op('restoreDoc', {'collection': collection, 'id': id});

  // --- Query ---
  Future<dynamic> findDocs(String collection, Map<String, dynamic> query) =>
      _op('findDocs', {'collection': collection, 'query': query});
  Future<dynamic> findDeletedDocs(
          String collection, Map<String, dynamic> query) =>
      _op('findDeletedDocs', {'collection': collection, 'query': query});
  Future<dynamic> findDocsPage(String collection, Map<String, dynamic> query,
          Map<String, dynamic> page) =>
      _op('findDocs', {'collection': collection, 'query': query, 'page': page});
  Future<dynamic> findDeletedDocsPage(String collection,
          Map<String, dynamic> query, Map<String, dynamic> page) =>
      _op('findDeletedDocs',
          {'collection': collection, 'query': query, 'page': page});
  Future<dynamic> joinDocs(Map<String, dynamic> join) =>
      _op('joinDocs', {'join': join});
  Future<dynamic> executeSQL(String sql, {int? uid, int? gid, int? mode}) =>
      _op('executeSQL', {
        'sql': sql,
        'access': uid == null && gid == null && mode == null
            ? null
            : {
                if (uid != null) 'uid': uid,
                if (gid != null) 'gid': gid,
                if (mode != null) 'mode': mode
              }
      });

  /// Run raw SQL, built with native interpolation: db.sql("… ${x}").
  /// Values are inlined verbatim — escape/validate untrusted input yourself.
  Future<dynamic> sql(String query, {int? uid, int? gid, int? mode}) =>
      executeSQL(query, uid: uid, gid: gid, mode: mode);

  Future<dynamic> importBulkData(String collection, String url) =>
      _op('importBulkData', {'collection': collection, 'url': url});

  /// Collection-scoped facade with short method names, so
  /// `db.collection('users').put(data)` reads like the browser client.
  FyloCollection collection(String name) => FyloCollection(this, name);

  /// Close stdin so the loop ends, and wait for the process to exit.
  Future<void> close() async {
    await _stdin.close();
    await _proc.exitCode;
  }
}

/// A collection-scoped view; methods drop the leading collection argument.
class FyloCollection {
  final Fylo _db;
  final String _name;

  FyloCollection(this._db, this._name);

  Future<dynamic> create([String kind = 'document']) =>
      _db.createCollection(_name, kind);
  Future<dynamic> drop() => _db.dropCollection(_name);
  Future<dynamic> inspect() => _db.inspectCollection(_name);
  Future<dynamic> rebuild() => _db.rebuildCollection(_name);
  Future<dynamic> put(Map<String, dynamic> data) => _db.putData(_name, data);
  Future<dynamic> get(String id) => _db.getDoc(_name, id);
  Future<dynamic> getMeta(String id) => _db.getMeta(_name, id);
  Future<dynamic> setMeta(String id, Map<String, dynamic> meta) =>
      _db.setMeta(_name, id, meta);
  Future<dynamic> latest(String id) => _db.getLatest(_name, id);
  Future<dynamic> patch(String id, Map<String, dynamic> newDoc) =>
      _db.patchDoc(_name, id, newDoc);
  Future<dynamic> delete(String id) => _db.delDoc(_name, id);
  Future<dynamic> restore(String id) => _db.restoreDoc(_name, id);
  Future<dynamic> find(Map<String, dynamic> query) =>
      _db.findDocs(_name, query);
  Future<dynamic> findPage(
          Map<String, dynamic> query, Map<String, dynamic> page) =>
      _db.findDocsPage(_name, query, page);
}

File: clients/go/fylo.go
// Package fylo drives the `fylo` binary's persistent NDJSON loop.
//
// Stdlib only. Requires the `fylo` binary on PATH (brew/scoop) or an explicit
// path. One long-lived subprocess keeps the engine warm across calls.
//
//	db, _ := fylo.Open("/path/to/db", "fylo")
//	defer db.Close()
//	db.CreateCollection("users", "document")
//	id, _ := db.PutData("users", map[string]any{"name": "Ada", "role": "admin"})
//	doc, _ := db.GetLatest("users", id.(string))
//	admins, _ := db.FindDocs("users", map[string]any{
//		"$ops": []any{map[string]any{"role": map[string]any{"$eq": "admin"}}}})
//
// Each operation method builds the request and returns the op's `result`
// (or an error if the op failed). Method names mirror the machine-protocol op
// names in Go's exported PascalCase. Request(op) remains a raw escape hatch
// returning the full response — use it for ops without a dedicated method.
package fylo

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"os/exec"
	"strconv"
	"sync"
)

const (
	maxRequestBytes  = 1024 * 1024
	maxResponseBytes = 8 * 1024 * 1024
)

type Fylo struct {
	cmd  *exec.Cmd
	pipe io.WriteCloser
	in   *bufio.Writer
	out  *bufio.Reader
	mu   sync.Mutex
}

// QueueConsumerOptions configures one bounded decorated invocation.
type QueueConsumerOptions struct {
	MaxMessages         int
	VisibilityTimeoutMs int
	MaxAttempts         int
	RetryDelayMs        int
}

// QueueProcessResult summarizes how a decorated batch was settled.
type QueueProcessResult struct {
	Claimed      int
	Acknowledged int
	Retried      int
	DeadLettered int
}

// QueueHandler processes one native delivery object.
type QueueHandler func(delivery map[string]any) error

// Open starts a warm fylo process rooted at root. binary defaults to "fylo".
func Open(root, binary string) (*Fylo, error) {
	if binary == "" {
		binary = "fylo"
	}
	args := []string{
		"exec", "--loop", "--root", root,
		"--max-request-bytes", strconv.Itoa(maxRequestBytes),
		"--max-response-bytes", strconv.Itoa(maxResponseBytes),
	}
	cmd := exec.Command(binary, args...)
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return nil, err
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, err
	}
	if err := cmd.Start(); err != nil {
		return nil, err
	}
	return &Fylo{
		cmd: cmd, pipe: stdin, in: bufio.NewWriter(stdin),
		out: bufio.NewReaderSize(stdout, maxResponseBytes+2),
	}, nil
}

// op builds a request, sends it, and returns `result` (or an error on failure).
func (f *Fylo) op(name string, fields map[string]any) (any, error) {
	payload := map[string]any{"op": name}
	for k, v := range fields {
		if v != nil {
			payload[k] = v
		}
	}
	resp, err := f.Request(payload)
	if err != nil {
		return nil, err
	}
	if ok, _ := resp["ok"].(bool); !ok {
		msg := "fylo error"
		if e, ok := resp["error"].(map[string]any); ok {
			if m, ok := e["message"].(string); ok {
				msg = m
			}
		}
		return nil, fmt.Errorf("%s", msg)
	}
	return resp["result"], nil
}

// --- Collections ---

func (f *Fylo) CreateCollection(collection, kind string) (any, error) {
	if kind == "" {
		kind = "document"
	}
	return f.op("createCollection", map[string]any{"collection": collection, "kind": kind})
}
func (f *Fylo) DropCollection(collection string) (any, error) {
	return f.op("dropCollection", map[string]any{"collection": collection})
}
func (f *Fylo) InspectCollection(collection string) (any, error) {
	return f.op("inspectCollection", map[string]any{"collection": collection})
}
func (f *Fylo) RebuildCollection(collection string) (any, error) {
	return f.op("rebuildCollection", map[string]any{"collection": collection})
}

// Durable serverless queue.
func (f *Fylo) QueuePublish(topic string, payload any, options map[string]any) (any, error) {
	fields := map[string]any{"topic": topic, "payload": payload}
	copyQueueOptions(fields, options, "delayMs", "idempotencyKey")
	return f.op("queuePublish", fields)
}
func (f *Fylo) QueueClaim(topic, group string, options map[string]any) (any, error) {
	fields := map[string]any{"topic": topic, "group": group}
	copyQueueOptions(fields, options, "maxMessages", "visibilityTimeoutMs", "maxAttempts")
	return f.op("queueClaim", fields)
}
func (f *Fylo) QueueAck(topic, group, id, receipt string) (any, error) {
	return f.op("queueAck", map[string]any{"topic": topic, "group": group, "id": id, "receipt": receipt})
}
func (f *Fylo) QueueNack(topic, group, id, receipt string, options map[string]any) (any, error) {
	fields := map[string]any{"topic": topic, "group": group, "id": id, "receipt": receipt}
	copyQueueOptions(fields, options, "delayMs", "reason")
	return f.op("queueNack", fields)
}
func (f *Fylo) QueueExtend(topic, group, id, receipt string, visibilityTimeoutMs int) (any, error) {
	return f.op("queueExtend", map[string]any{"topic": topic, "group": group, "id": id, "receipt": receipt, "visibilityTimeoutMs": visibilityTimeoutMs})
}
func (f *Fylo) QueueStats(topic, group string) (any, error) {
	return f.op("queueStats", map[string]any{"topic": topic, "group": group})
}
func (f *Fylo) QueueDeadLetters(topic, group string, limit int) (any, error) {
	return f.op("queueDeadLetters", map[string]any{"topic": topic, "group": group, "limit": limit})
}

// copyQueueOptions ignores unknown and protected request fields by construction.
func copyQueueOptions(fields, options map[string]any, allowed ...string) {
	for _, key := range allowed {
		if value, ok := options[key]; ok {
			fields[key] = value
		}
	}
}

// QueueProcess claims and settles one bounded batch. Handler errors are nacked.
func (f *Fylo) QueueProcess(topic, group string, handler QueueHandler, options QueueConsumerOptions) (QueueProcessResult, error) {
	if handler == nil {
		return QueueProcessResult{}, fmt.Errorf("queue handler is required")
	}
	if options.MaxMessages == 0 {
		options.MaxMessages = 1
	}
	if options.VisibilityTimeoutMs == 0 {
		options.VisibilityTimeoutMs = 30000
	}
	if options.MaxAttempts == 0 {
		options.MaxAttempts = 3
	}
	claimed, err := f.QueueClaim(topic, group, map[string]any{
		"maxMessages":         options.MaxMessages,
		"visibilityTimeoutMs": options.VisibilityTimeoutMs,
		"maxAttempts":         options.MaxAttempts,
	})
	if err != nil {
		return QueueProcessResult{}, err
	}
	deliveries, ok := claimed.([]any)
	if !ok {
		return QueueProcessResult{}, fmt.Errorf("fylo queue claim returned an invalid delivery list")
	}
	result := QueueProcessResult{Claimed: len(deliveries)}
	for _, item := range deliveries {
		delivery, ok := item.(map[string]any)
		if !ok {
			return result, fmt.Errorf("fylo queue claim returned an invalid delivery")
		}
		id, idOK := delivery["id"].(string)
		receipt, receiptOK := delivery["receipt"].(string)
		if !idOK || !receiptOK {
			return result, fmt.Errorf("fylo queue delivery lacks an id or receipt")
		}
		if handler(delivery) == nil {
			if _, err := f.QueueAck(topic, group, id, receipt); err != nil {
				return result, err
			}
			result.Acknowledged++
		} else {
			settled, err := f.QueueNack(topic, group, id, receipt, map[string]any{
				"delayMs": options.RetryDelayMs,
				"reason":  "queue handler failed",
			})
			if err != nil {
				return result, err
			}
			outcome, _ := settled.(map[string]any)
			if dead, _ := outcome["deadLettered"].(bool); dead {
				result.DeadLettered++
			} else {
				result.Retried++
			}
		}
	}
	return result, nil
}

// QueueConsumer returns Go's decorator-equivalent one-batch handler wrapper.
func (f *Fylo) QueueConsumer(topic, group string, handler QueueHandler, options QueueConsumerOptions) func() (QueueProcessResult, error) {
	return func() (QueueProcessResult, error) {
		return f.QueueProcess(topic, group, handler, options)
	}
}

// --- Documents ---

func (f *Fylo) PutData(collection string, data map[string]any) (any, error) {
	return f.op("putData", map[string]any{"collection": collection, "data": data})
}
func (f *Fylo) BatchPutData(collection string, batch []any) (any, error) {
	return f.op("batchPutData", map[string]any{"collection": collection, "batch": batch})
}
func (f *Fylo) GetDoc(collection, id string) (any, error) {
	return f.op("getDoc", map[string]any{"collection": collection, "id": id})
}
func (f *Fylo) GetMeta(collection, id string) (any, error) {
	return f.op("getMeta", map[string]any{"collection": collection, "id": id})
}
func (f *Fylo) SetMeta(collection, id string, meta map[string]any) (any, error) {
	return f.op("setMeta", map[string]any{"collection": collection, "id": id, "meta": meta})
}
func (f *Fylo) GetLatest(collection, id string) (any, error) {
	return f.op("getLatest", map[string]any{"collection": collection, "id": id})
}
func (f *Fylo) PatchDoc(collection, id string, newDoc map[string]any) (any, error) {
	return f.op("patchDoc", map[string]any{"collection": collection, "id": id, "newDoc": newDoc})
}
func (f *Fylo) PatchDocs(collection string, update map[string]any) (any, error) {
	return f.op("patchDocs", map[string]any{"collection": collection, "update": update})
}
func (f *Fylo) DelDoc(collection, id string) (any, error) {
	return f.op("delDoc", map[string]any{"collection": collection, "id": id})
}
func (f *Fylo) DelDocs(collection string, criteria map[string]any) (any, error) {
	return f.op("delDocs", map[string]any{"collection": collection, "delete": criteria})
}
func (f *Fylo) RestoreDoc(collection, id string) (any, error) {
	return f.op("restoreDoc", map[string]any{"collection": collection, "id": id})
}

// --- Query ---

func (f *Fylo) FindDocs(collection string, query map[string]any) (any, error) {
	return f.op("findDocs", map[string]any{"collection": collection, "query": query})
}
func (f *Fylo) FindDeletedDocs(collection string, query map[string]any) (any, error) {
	return f.op("findDeletedDocs", map[string]any{"collection": collection, "query": query})
}
func (f *Fylo) FindDocsPage(collection string, query, page map[string]any) (any, error) {
	return f.op("findDocs", map[string]any{"collection": collection, "query": query, "page": page})
}
func (f *Fylo) FindDeletedDocsPage(collection string, query, page map[string]any) (any, error) {
	return f.op("findDeletedDocs", map[string]any{"collection": collection, "query": query, "page": page})
}
func (f *Fylo) JoinDocs(join map[string]any) (any, error) {
	return f.op("joinDocs", map[string]any{"join": join})
}
func (f *Fylo) ExecuteSQL(sql string, access ...map[string]any) (any, error) {
	fields := map[string]any{"sql": sql}
	if len(access) > 0 {
		fields["access"] = access[0]
	}
	return f.op("executeSQL", fields)
}

// Sql runs raw SQL, built with fmt.Sprintf. Values are inlined verbatim —
// escape/validate untrusted input yourself.
func (f *Fylo) Sql(query string, access ...map[string]any) (any, error) {
	return f.ExecuteSQL(query, access...)
}
func (f *Fylo) ImportBulkData(collection, url string) (any, error) {
	return f.op("importBulkData", map[string]any{"collection": collection, "url": url})
}

// Request sends one raw machine-protocol op and returns the full response.
func (f *Fylo) Request(op map[string]any) (map[string]any, error) {
	line, err := json.Marshal(op)
	if err != nil {
		return nil, err
	}
	if len(line) > maxRequestBytes {
		return nil, fmt.Errorf("FYLO request exceeds %d bytes", maxRequestBytes)
	}
	f.mu.Lock() // ponytail: one call in flight; drop the lock only if you pipeline
	defer f.mu.Unlock()
	if _, err := f.in.Write(append(line, '\n')); err != nil {
		return nil, err
	}
	if err := f.in.Flush(); err != nil {
		return nil, err
	}
	reply, err := f.out.ReadSlice('\n')
	if err != nil || len(reply) == 0 || reply[len(reply)-1] != '\n' || len(reply)-1 > maxResponseBytes {
		_ = f.cmd.Process.Kill()
		if err != nil {
			return nil, fmt.Errorf("fylo response framing failed: %w", err)
		}
		return nil, fmt.Errorf("FYLO response exceeds %d bytes", maxResponseBytes)
	}
	var resp map[string]any
	if err := json.Unmarshal(reply, &resp); err != nil {
		_ = f.cmd.Process.Kill()
		return nil, err
	}
	return resp, nil
}

// Close ends the process by closing stdin and waiting for exit.
func (f *Fylo) Close() error {
	_ = f.pipe.Close()
	return f.cmd.Wait()
}

// Collection returns a collection-scoped facade with short method names, so
// db.Collection("users").Put(data) reads like the browser client.
func (f *Fylo) Collection(name string) *Collection {
	return &Collection{fylo: f, name: name}
}

// Collection is a collection-scoped view; methods drop the leading collection arg.
type Collection struct {
	fylo *Fylo
	name string
}

func (c *Collection) Create(kind string) (any, error) { return c.fylo.CreateCollection(c.name, kind) }
func (c *Collection) Drop() (any, error)              { return c.fylo.DropCollection(c.name) }
func (c *Collection) Inspect() (any, error)           { return c.fylo.InspectCollection(c.name) }
func (c *Collection) Rebuild() (any, error)           { return c.fylo.RebuildCollection(c.name) }
func (c *Collection) Put(data map[string]any) (any, error) {
	return c.fylo.PutData(c.name, data)
}
func (c *Collection) Get(id string) (any, error)     { return c.fylo.GetDoc(c.name, id) }
func (c *Collection) GetMeta(id string) (any, error) { return c.fylo.GetMeta(c.name, id) }
func (c *Collection) SetMeta(id string, meta map[string]any) (any, error) {
	return c.fylo.SetMeta(c.name, id, meta)
}
func (c *Collection) Latest(id string) (any, error) { return c.fylo.GetLatest(c.name, id) }
func (c *Collection) Patch(id string, newDoc map[string]any) (any, error) {
	return c.fylo.PatchDoc(c.name, id, newDoc)
}
func (c *Collection) Delete(id string) (any, error)  { return c.fylo.DelDoc(c.name, id) }
func (c *Collection) Restore(id string) (any, error) { return c.fylo.RestoreDoc(c.name, id) }
func (c *Collection) Find(query map[string]any) (any, error) {
	return c.fylo.FindDocs(c.name, query)
}
func (c *Collection) FindPage(query, page map[string]any) (any, error) {
	return c.fylo.FindDocsPage(c.name, query, page)
}
